File: packages/sdk-ts/src/resources/events.ts
import type { HttpClient } from '../http.js';

export interface GrantexEvent {
  id: string;
  type: string;
  createdAt: string;
  data: Record<string, unknown>;
}

export interface StreamOptions {
  types?: string[];
}

export type EventHandler = (event: GrantexEvent) => void | Promise<void>;

export class EventsClient {
  #http: HttpClient;

  constructor(http: HttpClient) {
    this.#http = http;
  }

  /**
   * Connect to the SSE event stream.
   * Returns an async iterable of events.
   */
  async *stream(options?: StreamOptions): AsyncGenerator<GrantexEvent> {
    const params = new URLSearchParams();
    if (options?.types?.length) {
      params.set('types', options.types.join(','));
    }
    const qs = params.toString();
    const url = `/v1/events/stream${qs ? '?' + qs : ''}`;

    const res = await this.#http.rawGet(url);
    if (!res.ok || !res.body) {
      throw new Error(`Failed to connect to event stream: ${res.status}`);
    }

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() ?? '';

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              yield JSON.parse(line.slice(6)) as GrantexEvent;
            } catch { /* skip malformed */ }
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Subscribe to events with a callback handler.
   * Returns a function to unsubscribe.
   */
  subscribe(handler: EventHandler, options?: StreamOptions): { unsubscribe: () => void } {
    const controller = new AbortController();
    const run = async () => {
      for await (const event of this.stream(options)) {
        if (controller.signal.aborted) break;
        await handler(event);
      }
    };
    run().catch(() => {});

    return {
      unsubscribe: () => controller.abort(),
    };
  }
}

File: packages/sdk-py/src/grantex/resources/_events.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Iterator, Optional, Sequence
import json
import threading

import httpx


EventHandler = Callable[["GrantexEvent"], None]
ErrorHandler = Callable[[Exception], None]


@dataclass(frozen=True)
class GrantexEvent:
    id: str
    type: str
    created_at: str
    data: dict[str, Any]

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> "GrantexEvent":
        return cls(
            id=d["id"],
            type=d["type"],
            created_at=d["createdAt"],
            data=d.get("data", {}),
        )


@dataclass(frozen=True)
class StreamOptions:
    types: Optional[Sequence[str]] = None


class Subscription:
    """Handle returned by ``EventsClient.subscribe`` to control the background stream."""

    def __init__(self, thread: threading.Thread, stop_event: threading.Event) -> None:
        self._thread = thread
        self._stop_event = stop_event

    @property
    def active(self) -> bool:
        """Return ``True`` if the subscription is still running."""
        return self._thread.is_alive()

    def unsubscribe(self) -> None:
        """Stop the background stream and wait for the thread to exit."""
        self._stop_event.set()
        self._thread.join(timeout=5)


class EventsClient:
    def __init__(self, base_url: str, api_key: str) -> None:
        self._base_url = base_url
        self._api_key = api_key

    def stream(self, options: Optional[StreamOptions] = None) -> Iterator[GrantexEvent]:
        """Connect to the SSE event stream. Yields GrantexEvent objects."""
        params = {}
        if options and options.types:
            params["types"] = ",".join(options.types)

        url = f"{self._base_url}/v1/events/stream"
        with httpx.stream(
            "GET",
            url,
            params=params,
            headers={"Authorization": f"Bearer {self._api_key}"},
            timeout=None,
        ) as response:
            response.raise_for_status()
            buffer = ""
            for chunk in response.iter_text():
                buffer += chunk
                while "\n" in buffer:
                    line, buffer = buffer.split("\n", 1)
                    if line.startswith("data: "):
                        try:
                            data = json.loads(line[6:])
                            yield GrantexEvent.from_dict(data)
                        except (json.JSONDecodeError, KeyError):
                            pass

    def subscribe(
        self,
        handler: EventHandler,
        options: Optional[StreamOptions] = None,
        *,
        on_error: Optional[ErrorHandler] = None,
    ) -> Subscription:
        """Subscribe to events with a callback handler.

        Starts a background thread that calls ``stream()`` and invokes
        *handler* for each event.  Returns a :class:`Subscription` whose
        ``unsubscribe()`` method stops the thread.

        Args:
            handler: Called with each :class:`GrantexEvent` received.
            options: Optional :class:`StreamOptions` to filter event types.
            on_error: Optional callback invoked when the stream raises an
                exception.  If not provided, errors are silently swallowed
                and the stream stops.
        """
        stop = threading.Event()

        def _run() -> None:
            try:
                for event in self.stream(options):
                    if stop.is_set():
                        break
                    handler(event)
            except Exception as exc:  # noqa: BLE001
                if on_error is not None:
                    on_error(exc)

        thread = threading.Thread(target=_run, daemon=True)
        thread.start()
        return Subscription(thread=thread, stop_event=stop)
